### ISSUE TITLE
`ArrayObject` false positive `expects ArrayObject<array-key, mixed>, but ArrayObject<string, mixed> provided`

### ISSUE THREAD

**Issue by boesing** [vimeo/psalm#10579]
https://psalm.dev/r/3de661563e

**Comment by weirdan** [comment]
@boesing mind elaborating on why you think it's a false positive? Keep in mind that `ArrayObject` allows writes, so its type parameters cannot be covariant.

**Comment by boesing** [comment]
@weirdan array-key is allowed, but when passing string its invalid. 
That makes no sense since array-key is an alias for string|int.

**Comment by weirdan** [comment]
Consider the following code:
```php
/** @param ArrayObject<array-key, mixed> $a */
function foo(ArrayObject $a): void {
  $a->offsetSet(20, 'dummy');
}

/** @var ArrayObject<string, mixed> $bar */
$bar = new ArrayObject(["a" => "b"]);
f($bar);

// what is the type of $bar now?
```

Also, see https://psalm.dev/docs/annotating_code/templated_annotations/#template-covariance

**Comment by boesing** [comment]
I understand that some1 can change the value. but I do not say its string. its still array-key.

I dont get why it has to cry before anything was passed somewhere.

I would argue that if passing an arrayobject to an immutable function, the type is preserved. if passing to a normal one it should be flagged.

To me, my issue is still not solved.
Instantiating an splqueue had the same issues. To me, it sounds like there is no reason to use php internal classes at all (i.e. SPL or ArrayObject, etc.) as they do not force specific types.
So most probably those classes should be considered bad practice due to their mutability, right?
